Algorithm/include/Algorithm/PageParser.h

```diff
@@ -0,0 +1,172 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef ALGORITHM_PAGEPARSER_H
+#define ALGORITHM_PAGEPARSER_H
+
+/// @file   PageParser.h
+/// @author Matthias Richter
+/// @since  2017-09-27
+/// @brief  Parser for a set of data objects in consecutive memory pages.
+
+#include <functional>
+#include <vector>
+
+namespace o2 {
+
+namespace algorithm {
+
+/**
+ * @class PageParser
+ * Parser for a set of data objects in consecutive memory pages.
+ *
+ * All memory pages have a fixed size and start with a page header.
+ * Depending on the page size and size of the data object, some
+ * objects can be split at the page boundary and have the page header
+ * embedded.
+ *
+ * The class iterator can be used to iterate over the data objects
+ * transparently.
+ *
+ * Usage:
+ *   RawParser<PageHeaderType, N, ElementType> RawParser;
+ *   RawParser parser(ptr, size);
+ *   for (auto element : parser) {
+ *     // do something with element
+ *   }
+ */
+template<typename PageHeaderT,
+         size_t PageSize,
+         typename ElementType,
+         typename GroupType = int // later extension to groups of elements
+         >
+class PageParser {
+public:
+  using PageHeaderType = PageHeaderT;
+  using BufferType = unsigned char;
+  using value_type = ElementType;
+  using group_type = GroupType;
+  using GetNElements = std::function<size_t(const group_type&)>;
+  static const size_t page_size = PageSize;
+
+  PageParser() = delete;
+  PageParser(const BufferType* buffer, size_t size,
+             GetNElements getNElementsFct = [] (const group_type&) {return 0;}
+             )
+    : mBuffer(buffer)
+    , mSize(size)
+    , mGetNElementsFct(getNElementsFct)
+    , mNPages(size>0?(size/page_size)+1:0)
+  {
+  }
+  ~PageParser() = default;
+
+  using IteratorBase = std::iterator<std::forward_iterator_tag, value_type>;
+
+  class iterator : public IteratorBase {
```

Comment:
shouldn't types start with a capital letter?

Reply:
true, but as it is it's more in the style of std::iterator

Reply:
Not sure about this. An iterator is defined via operators not via a standard name. In any case, such exceptions will make automatic code checks very hard (to be reliable) and probably we would need to introduce a way to annotate rule violations.